### PR TITLE
feat: change password + refactor Field component

### LINKED ISSUE #61: Basic User Settings
## Profile

- [x] Custom chat color
- [ ] Profile Picture (Depends on #123)
- [x] Display name (capitalization of username)

## Account

- [x] Change password (#141)
- [x] Change email
- [x] 2FA settings (#125)

This issue doesn't contain more advanced settings like notification and billing settings.

### PR DESCRIPTION
**Pull request by lennartkloock**
## Proposed changes

- Adds change password button in account settings and endpoint.
- Refactor `Field` component

## Types of changes

What types of changes does your code introduce to Scuffle?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have read the [CONTRIBUTING](https://github.com/ScuffleTV/scuffle/blob/main/CONTRIBUTING.md) doc
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

Closes #61 and #132

### CODE REVIEW

File: platform/api/src/api/v1/gql/mutations/user.rs [hunk 1]
```diff
@@ -148,6 +149,59 @@ impl UserMutation {
         Ok(user.into())
     }
 
+    async fn password<'ctx>(
+        &self,
+        ctx: &Context<'_>,
+        #[graphql(desc = "Current password")] current_password: String,
+        #[graphql(desc = "New password", validator(custom = "PasswordValidator"))]
+        new_password: String,
+    ) -> Result<User> {
```
Comment (TroyKomodo) [inline platform/api/src/api/v1/gql/mutations/user.rs:158]: What happens with 2FA?

Comment (lennartkloock) [reply platform/api/src/api/v1/gql/mutations/user.rs:158]: It stays like before. Or what do you mean?

File: platform/website/src/components/auth/auth-dialog.svelte [hunk 2]
```diff
@@ -48,56 +50,38 @@
...
-		},
-		reload() {
-			username.touched = username.value.length > 0;
+	$: $authDialog, resetAllFields();
 
+	let usernameValue: string;
+	let usernameStatus: FieldStatus;
+	let usernameValidationTimeout: number | NodeJS.Timeout;
+	function usernameValidate(v: string): Promise<FieldStatus> {
+		return new Promise((resolve) => {
```
Comment (TroyKomodo) [inline platform/website/src/components/auth/auth-dialog.svelte:59]: Can this be an async function?

File: platform/website/src/components/auth/auth-dialog.svelte [hunk 3]
```diff
@@ -109,201 +93,58 @@
...
+	}
 
-			if (password.value !== confirmPassword.value) {
-				confirmPassword.status = "error";
-				confirmPassword.message = "Passwords do not match";
-				return;
-			}
+	let emailStatus: FieldStatus;
+	let emailValue: string;
+	async function emailValidate(v: string): Promise<FieldStatus> {
```
Comment (TroyKomodo) [inline platform/website/src/components/auth/auth-dialog.svelte:118]: Do async functions need to annotate that their resp type is Promise in TypeScript?

File: platform/website/src/components/settings/account/change-password.svelte [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+		);
+	}
+
+	async function changePassword() {
+		if (formValid && currentPassword && newPassword) {
+			loading = true;
+			const res = await client
+				.mutation(
+					graphql(`
+						mutation ChangePassword($currentPassword: String!, $newPassword: String!) {
```
Comment (TroyKomodo) [inline platform/website/src/components/settings/account/change-password.svelte:48]: is this formatted?

**Review by TroyKomodo — COMMENTED** [review]
Looks really good!

File: platform/website/src/components/settings/account/change-password.svelte (hunk 4 above)
Comment (lennartkloock) [reply platform/website/src/components/settings/account/change-password.svelte:48]: Yes, it is

File: platform/website/src/components/auth/auth-dialog.svelte (hunk 3 above)
Comment (lennartkloock) [reply platform/website/src/components/auth/auth-dialog.svelte:118]: No, no idea why I did that LULE

File: platform/website/src/components/auth/auth-dialog.svelte (hunk 2 above)
Comment (lennartkloock) [reply platform/website/src/components/auth/auth-dialog.svelte:59]: I want to call the `resolve` function in the `setTimeout` callback, so I explicitly want to create the Promise. Let me know if there is a better way of doing it.

Comment (TroyKomodo) [reply platform/website/src/components/auth/auth-dialog.svelte:59]: Another way would be to wrap the set timeout itself in a promise, would make things look nicer I think.

Comment (lennartkloock) [reply platform/website/src/components/auth/auth-dialog.svelte:59]: What do you mean? I want to resolve the usernameValidate Promise in the setTimeout callback

File: platform/website/src/components/auth/auth-dialog.svelte [hunk 5]
```diff
@@ -48,56 +50,38 @@
...
-			// We clear the timeout so that we dont send a request to the server on every character change.
-			clearTimeout(username.extra.timeout);
-
-			// When we are in login mode, we dont need to validate the username.
-			if ($authDialog !== AuthDialog.Register) {
+				resolve({ type: FieldStatusType.Success });
 				return;
 			}
 
+			clearTimeout(usernameValidationTimeout);
```
Comment (TroyKomodo) [inline platform/website/src/components/auth/auth-dialog.svelte:66]: Does this cause a memory leak?
What happens when you drop this if the promise has never resolved?

File: platform/website/src/components/auth/auth-dialog.svelte (hunk 2 above)
Comment (TroyKomodo) [reply platform/website/src/components/auth/auth-dialog.svelte:59]: You wrap the settimeout in a promise, and then await it

Comment (lennartkloock) [reply platform/website/src/components/auth/auth-dialog.svelte:59]: Sorry, I still don't really get how this would work. Could you suggest some code?

Comment (TroyKomodo) [reply platform/website/src/components/auth/auth-dialog.svelte:59]:
```ts
async function xyz(): string {
    await new Promise(resolve => setTimeout(() => resolve(), 2000));
    return "hi from 2s later";
}
```

Comment (lennartkloock) [reply platform/website/src/components/auth/auth-dialog.svelte:59]: oh yeah like this, makes sense

Comment (lennartkloock) [reply platform/website/src/components/auth/auth-dialog.svelte:59]: But how do we cancel it then?

Comment (lennartkloock) [reply platform/website/src/components/auth/auth-dialog.svelte:59]: nvm, it seems like it just magically works

File: platform/website/src/components/auth/auth-dialog.svelte (hunk 5 above)
Comment (lennartkloock) [reply platform/website/src/components/auth/auth-dialog.svelte]: I think this is solved by the refactoring. Is it? I'm not sure

File: platform/website/src/components/auth/auth-dialog.svelte [hunk 6]
```diff
@@ -48,262 +50,99 @@
...
-										id
-									}
-								}
+		usernameStatus = {
+			type: FieldStatusType.Loading,
+			message: "Hold on while we check if this username is available...",
+		};
+
+		await new Promise((resolve) => {
+			usernameValidationTimeout = setTimeout(resolve, 500);
```
Comment (TroyKomodo) [inline platform/website/src/components/auth/auth-dialog.svelte:82]: I am not well versed in what happens if we cancel the timeout here, perhaps we need to also reject the promise?

Comment (lennartkloock) [reply platform/website/src/components/auth/auth-dialog.svelte:82]: Actually, I think what happens is that the Promise never gets resolved and is waiting forever

Comment (TroyKomodo) [reply platform/website/src/components/auth/auth-dialog.svelte:82]: so it leaks memory?

Comment (lennartkloock) [reply platform/website/src/components/auth/auth-dialog.svelte:82]: probably yeah

Comment (TroyKomodo) [reply platform/website/src/components/auth/auth-dialog.svelte:82]: javascript 4Head

Comment (lennartkloock) [reply platform/website/src/components/auth/auth-dialog.svelte:82]: 4Kopf

Comment (TroyKomodo) [inline platform/website/src/components/auth/auth-dialog.svelte]: We still need to fix this, by exposing the `reject` part of the promise outside, and making it so when you cancel the timeout the reject gets called.

**Review by TroyKomodo — APPROVED** [review]

**Pull request merged**